Change schedulers to chainable form

vincentqb opened the pull request:
Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208). The commit is self-contained and includes #21800.

Stack from [ghstack](https://github.com/ezyang/ghstack):

* Changing the behavior of schedulers to their chainable formula when supported
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch

vincentqb committed:
* Changing the behavior of schedulers to the chainable formula
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler
* Returning a deprecation warning when invoking the undocumented get_lr function (see comment) referring to get_computed_values, and deprecating it in the next release.
* CosineAnnealingWarmRestart still take epoch parameter

ezyang commented:
Can you rebase and resubmit this?

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +78,26 @@ def load_state_dict(self, state_dict):
+        warnings.warn(
+            "get_computed_values is now the supported method to obtain the last "
+            "computed learning rate. get_lr now returns the same, but will be "
+            "deprecated in favor of get_computed_values.", DeprecationWarning
```
When I'm told that "instead of using `f`, I should use `g`" where they have exactly the same meaning, I tend to also like to know why a renaming happened at all. Is there actually a subtle semantics change? Is the old name bad and the new name more correctly says what happens? Just saying "foo is the new way of doing bar, bar returns the same thing" sounds like you've just been moving deck chairs around.

vincentqb replied:
Good point. I'll add more details, like
```
      "The behavior of schedulers has been changed from their closed form to their "                                                                                                                                                     
      "chainable form. As such, the epoch parameter no longer applies, and is being "                                                                                                                                                    
      "deprecated. During the deprecation, if epoch is different from None, the closed "                                                                                                                                                 
      "form is used whenever available instead of the chainable form." 
```

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -1,23 +1,40 @@
 
+EPOCH_DEPRECATION_WARNING = ("The epoch parameter is being deprecated. If epoch "
+                             "is different from None, the closed form is used "
+                             "whenever available instead of the chainable form.")
```
Yeah, I'm going to harp on deprecation warnings some more. Put yourself in the shoes of the user: they had some perfectly reasonable, working code on a previous version of PyTorch, and now it's throwing these warnings. What does the warning say? "Epoch is deprecated." OK. Why is it being deprecated? What should I use instead? "If epoch is different from None, the closed form is used whenever available instead of the chainable form." OK.... you didn't answer either of my questions!

vincentqb replied:
Thanks again for the feedback. I'll add the "why?" and the "then what?" like so
```
      "The behavior of schedulers has been changed from their closed form to their "                                                                                                                                                     
      "chainable form. As such, the epoch parameter no longer applies, and is being "                                                                                                                                                    
      "deprecated. The new approach is to invoke step without the epoch parameter "                                                                                                                                                      
      "during the iteration. During the deprecation, if epoch is different from None, "                                                                                                                                                  
      " the closed form is used whenever available instead of the chainable form."  
```

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -1,23 +1,40 @@
                 group.setdefault('initial_lr', group['lr'])
-            last_epoch = 0
+            # last_epoch = 0
+            # step_last_epoch = None
```
Kill these lines? (Unless you have a reason to keep them?)

vincentqb replied:
Thanks for pointing this out: I'll remove them.

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +78,26 @@ def load_state_dict(self, state_dict):
+        return self._last_computed_values
+
+    def _compute_values(self):
+        # NOTE [LR Chainable Form] Compute learning rate using chainable form of the scheduler
```
So, the way the `Note [Foo]` convention works in the codebase, is that you have some long explanation that lives somewhere else, like:

```
Note [Foo]
~~~~~~~~~
blah blah blah
```

And then you reference the note from relevant code fragments `See Note [Foo]`. I don't see any such long explanation here. Do you want to add a longer explanation? Otherwise, just get rid of the `Note`

vincentqb replied:
Thanks for the clarification. I'll keep the comment and remove the tag as you suggest.

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -26,6 +43,7 @@ def __init__(self, optimizer, last_epoch=-1):
         self.base_lrs = list(map(lambda group: group['initial_lr'], optimizer.param_groups))
         self.last_epoch = last_epoch
```
Previously, it was not possible for `self.last_epoch` to be set to `-1`. With your changes, it is now set to `-1` when you pass in the default. This change to the invariants of members of LRScheduler was not documented in the PR. Can you explain more to me why this is sound? Because the `last_epoch` field is not underscored, it counts as a public member field, and so this is a very clear BC-breaking change.

vincentqb replied:
The default value of `last_epoch` has not changed: it was and is `-1`. However, during the initialization, 

* now: `last_epoch` defaults to -1, `step` is called and increments `last_epoch` to 0. 
* before: `last_epoch` defaults to -1, changed to 0, and `step` is called and increments to 1.

All the tests though were calling `step` outside of initialization with the epoch parameter, thus overriding `last_epoch`. This indeed raises the question: is this change backward compatible? 

However, running test/test_optim.py with pytorch-nightly works, but leads to 5 test failures when removing `epoch` from `step` when you would expect the same answers as before.

```
diff --git a/test/test_optim.py b/test/test_optim.py
index e334962c9..2a365649a 100644
--- a/test/test_optim.py
+++ b/test/test_optim.py
@@ -565,7 +565,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -599,7 +599,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -627,7 +627,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -1153,7 +1153,7 @@ class TestLRScheduler(TestCase):
         if isinstance(schedulers, _LRScheduler):
             schedulers = [schedulers]
         for epoch in range(epochs):
-            [scheduler.step(epoch) for scheduler in schedulers]
+            [scheduler.step() for scheduler in schedulers]
             for param_group, target in zip(self.opt.param_groups, targets):
                 self.assertAlmostEqual(target[epoch], param_group['lr'],
                                        msg='LR is wrong in epoch {}: expected {}, got {}'.format(
@@ -1184,7 +1184,7 @@ class TestLRScheduler(TestCase):
             targets.append([group['lr'] for group in self.opt.param_groups])
         self.setUp()
         for epoch in range(epochs):
-            scheduler.step(epoch)
+            scheduler.step()
             for i, param_group in enumerate(self.opt.param_groups):
                 self.assertAlmostEqual(targets[epoch][i], param_group['lr'],
                                        msg='LR is wrong in epoch {}: expected {}, got {}'.format(
@@ -1198,7 +1198,7 @@ class TestLRScheduler(TestCase):
                 if isinstance(scheduler, ReduceLROnPlateau):
                     scheduler.step(metrics[epoch])
                 else:
-                    scheduler.step(epoch)
+                    scheduler.step()
             if verbose:
                 print('epoch{}:\tlr={}'.format(epoch, self.opt.param_groups[0]['lr']))
             for param_group, target in zip(self.opt.param_groups, targets):
```

I would in fact argue that this fixes the initial stepping behavior. Thoughts?

ezyang commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +78,26 @@ def load_state_dict(self, state_dict):
-    def get_lr(self):
+    def get_computed_values(self):
+        # Return last computed learning rate by current scheduler
+        return self._last_computed_values
```
So for example, a way `get_computed_values` differs from `get_lr` is that it's not valid to call `get_computed_values` on a LR scheduler which has not been stepped once. Maybe it would be safer and more obvious to our users if we raised an error in this case, instead of returning a list of NaNs (which may not be detected until much later when you're looking at your parameters and trying to figure out why they've all become NaNs...)

vincentqb replied:
Yes, let's return an error instead if step has not been run.

ezyang reviewed (changes requested):
I want to accept this. I really do. But the patch as written just seems too risky given that the branch cut is happening in a few days. cc'ing @gchanan and @soumith for a second opinion.

vincentqb committed:
Update on "Change schedulers to chainable form"

Implement the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208). The commit is self-contained and includes #21800.

Stack from [ghstack](https://github.com/ezyang/ghstack):

* Changing the behavior of schedulers to the chainable formula
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler
* Returning a deprecation warning when invoking the undocumented get_lr function (see comment) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch

vincentqb commented on test/test_optim.py:
```diff
@@ -1028,6 +1181,8 @@ def test_CosineAnnealingWarmRestarts_lr1(self):
                     T_i = int(T_mult) * T_i
                 targets[0] += [eta_min + (0.05 - eta_min) * (1 + math.cos(math.pi * T_cur / T_i)) / 2]
                 targets[1] += [eta_min + (0.5 - eta_min) * (1 + math.cos(math.pi * T_cur / T_i)) / 2]
+            # NOTE CosineAnnealingWarmRestarts requires step before
```
See fix mentioned [here](https://github.com/pytorch/pytorch/pull/23833#pullrequestreview-271629279).

vincentqb commented:
> Can you rebase and resubmit this?

Rebased in #24352
